File: todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/addedittask/AddEditTaskModule.java
package com.example.android.architecture.blueprints.todoapp.addedittask;

import android.app.Activity;
import android.content.Context;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.example.android.architecture.blueprints.todoapp.Injection;
import com.example.android.architecture.blueprints.todoapp.util.providers.BaseNavigator;

/**
 * Enables inversion of control of the ViewModel and Navigator classes for adding and editing tasks.
 */
class AddEditTaskModule {

    @NonNull
    public static AddEditTaskViewModel createAddEditTaskViewModel(@Nullable String taskId,
                                                                  @NonNull Activity activity) {
        Context appContext = activity.getApplicationContext();
        BaseNavigator navigationProvider = Injection.createNavigationProvider(activity);
        return new AddEditTaskViewModel(taskId, Injection.provideTasksRepository(appContext),
                createAddEditTaskNavigator(navigationProvider));
    }

    @NonNull
    public static AddEditTaskNavigator createAddEditTaskNavigator(
            @NonNull BaseNavigator navigationProvider) {
        return new AddEditTaskNavigator(navigationProvider);
    }

}

File: app/src/main/java/com/example/android/architecture/blueprints/todoapp/addedittask/di/AddEditTaskModule.kt
package com.example.android.architecture.blueprints.todoapp.addedittask.di

import androidx.lifecycle.ViewModel
import com.example.android.architecture.blueprints.todoapp.addedittask.AddEditTaskViewModel
import com.example.android.architecture.blueprints.todoapp.di.ViewModelKey
import dagger.Binds
import dagger.Module
import dagger.multibindings.IntoMap

@Module
abstract class AddEditTaskModule {

    @Binds
    @IntoMap
    @ViewModelKey(AddEditTaskViewModel::class)
    abstract fun bindViewModel(viewmodel: AddEditTaskViewModel): ViewModel
}
